src/Crate/PDO/Http/ServerPool.php

```diff
@@ -0,0 +1,227 @@
+<?php
+/**
+ * Licensed to CRATE Technology GmbH("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+namespace Crate\PDO\Http;
+
+use Crate\PDO\Exception\RuntimeException;
+use Crate\PDO\PDO;
+use Crate\Stdlib\Collection;
+use Crate\Stdlib\CollectionInterface;
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\BadResponseException;
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\RequestOptions;
+
+final class ServerPool implements ServerInterface
+{
+    private const DEFAULT_SERVER = 'localhost:4200';
+
+    /**
+     * @var string[]
+     */
+    private $availableServers  = [];
+    private $lastVisitedServer = 0;
+
+    /**
+     * @var Client
+     */
+    private $httpClient;
+
+    /**
+     * @var PDO
+     */
+    private $pdo;
+
+    /**
+     * Client constructor.
+     *
+     * @param PDO   $pdo
+     * @param array $servers
+     */
+    public function __construct(PDO $pdo, array $servers)
+    {
+        if (\count($servers) === 0) {
+            $servers = [self::DEFAULT_SERVER];
+        }
+
+        // micro optimization so we don't always hit the same server first
+        shuffle($servers);
+
+        foreach ($servers as $server) {
+            $this->availableServers[] = $server;
+        }
+
+        $this->pdo        = $pdo;
+        $this->httpClient = new Client();
+    }
+
+    /**
+     * {@Inheritdoc}
+     * @throws \GuzzleHttp\Exception\ConnectException
+     */
+    public function execute($query, array $parameters): CollectionInterface
+    {
+        $numServers = count($this->availableServers) - 1;
+
+        for ($i = 0; $i <= $numServers; $i++) {
+
+            // always get the first available server
+            $server = $this->availableServers[0];
+
+            // Move the selected server to the end of the stack
+            $this->availableServers[] = array_shift($this->availableServers);
+
+            $options = $this->getHttpOptions($server, $query, $parameters);
```

Comment:
@matriv I don't agree with that. Because all the changes are done via setting pdo attributes and not passed via the constructor. Recreating the server pool with new args because someone changed a setting seems a bit excessive. 

If we don't want to build the options array every time we execute a query then we could instead from the PDO class force the options to be rebuilt every time someone calls setAttribute. But then someone like @Ocramius comes a long and does some black magic reflection that breaks everything.

Reply:
@macnibblet Building the options on setAttribute seems better than doing it every time even if no attribute has changed

Reply:
@matriv I updated the code, it kinda got a bit messier tho.